src/java.base/share/classes/jdk/internal/misc/MainMethodFinder.java

```diff
@@ -0,0 +1,164 @@
+        return method != null && Modifier.isStatic(method.getModifiers());
+    }
+
+    private static boolean correctArgs(Method method) {
+        int argc = method.getParameterCount();
+
+        return argc == 0 || argc == 1 && method.getParameterTypes()[0] == String[].class;
+    }
+
+    /**
+     * Gather all the "main" methods in the class heirarchy.
+     *
+     * @param declc  the top level declaring class
+     * @param refc   the declaring class or super class
+     * @param mains  accumulated main methods
+     */
+    private static void gatherMains(Class<?> declc, Class<?> refc, List<Method> mains) {
+        if (refc != null && refc != Object.class) {
+            gatherMains(declc, refc.getSuperclass(), mains);
+
+            for (Method method : refc.getDeclaredMethods()) {
+                // Must be named "main", public|protected|package-private and either
+                // no arguments or one string array argument.
+                if ("main".equals(method.getName()) &&
+                        !isPrivate(method) &&
+                        correctArgs(method) &&
+                        // Only statics in the declaring class
+                        (!isStatic(method) || declc == refc)
+                ) {
+                    mains.add(method);
+                }
+            }
+        }
+    }
+
+    /**
+     * Comparator for two methods.
+     * Priority order is;
+     * static < non-static,
+     * public < non-public,
+     * string arg < no arg and
+     * sub-class < super-class.
+     *
+     * @param a  first method
+     * @param b  second method
+     *
+     * @return -1, 0 or 1 to represent higher priority. equals priority or lesser priority.
+     */
+    private static int compareMethods(Method a, Method b) {
+        int aMods = a.getModifiers();
+        int bMods = b.getModifiers();
+        boolean aIsStatic = Modifier.isStatic(aMods);
+        boolean bIsStatic = Modifier.isStatic(bMods);
+
+        if (aIsStatic && !bIsStatic) {
+            return -1;
+        } else if (bIsStatic && !aIsStatic) {
+            return 1;
+        }
+
+        boolean aIsPublic = Modifier.isPublic(aMods);
+        boolean bIsPublic = Modifier.isPublic(bMods);
+
+        if (aIsPublic && !bIsPublic) {
+            return -1;
+        } else if (bIsPublic && !aIsPublic) {
+            return 1;
+        }
+
+        int aCount = a.getParameterCount();
+        int bCount = b.getParameterCount();
+
+        if (aCount > bCount) {
+            return -1;
+        } else if (bCount > aCount) {
+            return 1;
+        }
+
+        Class<?> aClass = a.getDeclaringClass();
+        Class<?> bClass = b.getDeclaringClass();
+
+        if (bClass.isAssignableFrom(aClass)) {
+            return -1;
+        } else if (bClass.isAssignableFrom(aClass)) {
+            return 1;
+        }
+
+        return 0;
+    }
+
+    /**
+     * {@return priority main method or null if none found}
```

Comment:
I think this javadoc is perhaps outdated? The current implementation of this method throws a `NoSuchMethodException` when it can't find any eligible main method.

Reply:
Changing